File: app/src/main/java/app/android/ttp/mikazuki/yoshinani/binding/Converters.java
package app.android.ttp.mikazuki.yoshinani.binding;

import android.databinding.BindingAdapter;
import android.databinding.BindingConversion;
import android.support.annotation.NonNull;
import android.support.v4.util.Pair;
import android.view.View;
import android.widget.EditText;
import android.widget.RadioButton;
import android.widget.RadioGroup;
import android.widget.TextView;


import java.util.Calendar;

import app.android.ttp.mikazuki.yoshinani.R;
import app.android.ttp.mikazuki.yoshinani.model.TextWatcherAdapter;
import app.android.ttp.mikazuki.yoshinani.utils.ModelUtils;
import app.android.ttp.mikazuki.yoshinani.utils.TextUtils;

/**
 * @author haijimakazuki
 */
public class Converters {
    @BindingConversion
    public static String convertBindableToString(BindableString bindableString) {
        return bindableString.get();
    }

    @NonNull
    @BindingConversion
    public static String convertBindableToString(BindableInt bindableInt) {
        return String.valueOf(bindableInt.get());
    }

    @BindingConversion
    public static boolean convertBindableToBoolean(BindableBoolean bindableBoolean) {
        return bindableBoolean.get();
    }

    @BindingConversion
    public static String convertDate(Calendar date) {
        return ModelUtils.formatDate(date);
    }

    @BindingAdapter({"android:text"})
    public static void bindEditText(EditText view, final BindableString bindableString) {
        Pair<BindableString, TextWatcherAdapter> pair = (Pair) view.getTag(R.id.bound_observable);
        if (pair == null || pair.first != bindableString) {
            if (pair != null) {
                view.removeTextChangedListener(pair.second);
            }
            TextWatcherAdapter watcher = new TextWatcherAdapter() {
                @Override
                public void onTextChanged(CharSequence s, int start, int before, int count) {
                    bindableString.set(s.toString());
                }
            };
            view.setTag(R.id.bound_observable, new Pair<>(bindableString, watcher));
            view.addTextChangedListener(watcher);
        }
        String newValue = bindableString.get();
        if (!view.getText().toString().equals(newValue)) {
            view.setText(newValue);
        }
    }

    @BindingAdapter({"android:text"})
    public static void bindEditText(EditText view, final BindableInt bindableInt) {
        Pair<BindableInt, TextWatcherAdapter> pair = (Pair) view.getTag(R.id.bound_observable);
        if (pair == null || pair.first != bindableInt) {
            if (pair != null) {
                view.removeTextChangedListener(pair.second);
            }
            TextWatcherAdapter watcher = new TextWatcherAdapter() {
                @Override
                public void onTextChanged(CharSequence s, int start, int before, int count) {
                    if (android.text.TextUtils.isDigitsOnly(s.toString())) {
                        bindableInt.set(Integer.parseInt(s.toString()));
                    }
                }
            };
            view.setTag(R.id.bound_observable, new Pair<>(bindableInt, watcher));
            view.addTextChangedListener(watcher);
        }
        String newValue = String.valueOf(bindableInt.get());
        if (!view.getText().toString().equals(newValue)) {
            view.setText(newValue);
        }
    }

    @BindingAdapter({"android:text"})
    public static void bindRadioGroup(RadioGroup view, final BindableBoolean bindableBoolean) {
        if (view.getTag(R.id.bound_observable) != bindableBoolean) {
            view.setTag(R.id.bound_observable, bindableBoolean);
            view.setOnCheckedChangeListener((group, checkedId) -> bindableBoolean.set(checkedId == group.getChildAt(1).getId()));
        }
        Boolean newValue = bindableBoolean.get();
        ((RadioButton) view.getChildAt(newValue ? 1 : 0)).setChecked(true);
    }

    @BindingAdapter({"android:text"})
    public static void bindOnClick(View view, final Runnable runnable) {
        view.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                runnable.run();
            }
        });
    }

    @BindingAdapter("android:currency")
    public static void bindCurrency(TextView view, final double amount) {
        view.setText(TextUtils.wrapCurrency(amount));
    }
}

File: app/src/main/java/app/android/ttp/mikazuki/yoshinani/binding/Converters.kt
package app.android.ttp.mikazuki.yoshinani.binding

import android.databinding.BindingAdapter
import android.databinding.BindingConversion
import android.support.v4.util.Pair
import android.view.View
import android.widget.EditText
import android.widget.RadioButton
import android.widget.RadioGroup
import android.widget.TextView


import java.util.Calendar

import app.android.ttp.mikazuki.yoshinani.R
import app.android.ttp.mikazuki.yoshinani.model.TextWatcherAdapter
import app.android.ttp.mikazuki.yoshinani.utils.ModelUtils
import app.android.ttp.mikazuki.yoshinani.utils.TextUtils

/**
 * @author haijimakazuki
 */
object Converters {
    @BindingConversion
    fun convertBindableToString(bindableString: BindableString): String {
        return bindableString.get()
    }

    @BindingConversion
    fun convertBindableToString(bindableInt: BindableInt): String {
        return bindableInt.get().toString()
    }

    @BindingConversion
    fun convertBindableToBoolean(bindableBoolean: BindableBoolean): Boolean {
        return bindableBoolean.get()
    }

    @BindingConversion
    fun convertDate(date: Calendar): String {
        return ModelUtils.formatDate(date)
    }

    @BindingAdapter("android:text")
    fun bindEditText(view: EditText, bindableString: BindableString) {
        val pair = view.getTag(R.id.bound_observable) as Pair<*, *>
        if (pair == null || pair.first !== bindableString) {
            if (pair != null) {
                view.removeTextChangedListener(pair.second)
            }
            val watcher = object : TextWatcherAdapter() {
                override fun onTextChanged(s: CharSequence, start: Int, before: Int, count: Int) {
                    bindableString.set(s.toString())
                }
            }
            view.setTag(R.id.bound_observable, Pair<BindableString, TextWatcherAdapter>(bindableString, watcher))
            view.addTextChangedListener(watcher)
        }
        val newValue = bindableString.get()
        if (view.text.toString() != newValue) {
            view.setText(newValue)
        }
    }

    @BindingAdapter("android:text")
    fun bindEditText(view: EditText, bindableInt: BindableInt) {
        val pair = view.getTag(R.id.bound_observable) as Pair<*, *>
        if (pair == null || pair.first !== bindableInt) {
            if (pair != null) {
                view.removeTextChangedListener(pair.second)
            }
            val watcher = object : TextWatcherAdapter() {
                override fun onTextChanged(s: CharSequence, start: Int, before: Int, count: Int) {
                    if (android.text.TextUtils.isDigitsOnly(s.toString())) {
                        bindableInt.set(Integer.parseInt(s.toString()))
                    }
                }
            }
            view.setTag(R.id.bound_observable, Pair<BindableInt, TextWatcherAdapter>(bindableInt, watcher))
            view.addTextChangedListener(watcher)
        }
        val newValue = bindableInt.get().toString()
        if (view.text.toString() != newValue) {
            view.setText(newValue)
        }
    }

    @BindingAdapter("android:text")
    fun bindRadioGroup(view: RadioGroup, bindableBoolean: BindableBoolean) {
        if (view.getTag(R.id.bound_observable) !== bindableBoolean) {
            view.setTag(R.id.bound_observable, bindableBoolean)
            view.setOnCheckedChangeListener { group, checkedId -> bindableBoolean.set(checkedId == group.getChildAt(1).id) }
        }
        val newValue = bindableBoolean.get()
        (view.getChildAt(if (newValue) 1 else 0) as RadioButton).isChecked = true
    }

    @BindingAdapter("android:text")
    fun bindOnClick(view: View, runnable: Runnable) {
        view.setOnClickListener { runnable.run() }
    }

    @BindingAdapter("android:currency")
    fun bindCurrency(view: TextView, amount: Double) {
        view.text = TextUtils.wrapCurrency(amount)
    }
}